diff --git a/haskell/cabal.bzl b/haskell/cabal.bzl
index 02b77f61..aacf1b85 100644
--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -1013,8 +1013,8 @@ _STACK_BINDISTS = \
     {
         "2.7.5": {
             "linux-x86_64": (
-                "https://github.com/commercialhaskell/stack/releases/download/v2.7.5/stack-2.7.5-linux-x86_64-static.tar.gz",
-                "2a02fefefcc1758033d0aea566a521a290e3c68739ce9894bd6492a346af79c5",
+                "https://github.com/avdv/stack/releases/download/S-5170/stack-Linux.zip",
+                "da2e0ee35c9bed9ac42dc0c6a842c669e941edb31f31023a371c81fda5d1d327",
             ),
             "linux-aarch64": (
                 # GHCup provides unofficial aarch64 linux binaries
@@ -1023,8 +1023,8 @@ _STACK_BINDISTS = \
                 "f362fa4786b17252004b2619ec96b9687e561dc4e55c2612c53d60be767cabba",
             ),
             "osx-x86_64": (
-                "https://github.com/commercialhaskell/stack/releases/download/v2.7.5/stack-2.7.5-osx-x86_64.tar.gz",
-                "94176b71425d76b94b088515103316ae1ff96d123344b1f4609c103d0d5bdcc4",
+                "https://github.com/avdv/stack/releases/download/S-5170/stack-macOS.zip",
+                "232f8a31e0b93bf36b3d32a6c32a03407cb75e271dbf220451623393e943ff5e",
             ),
             "osx-aarch64": (
                 # GHCup provides unofficial aarch64 macOS binaries
@@ -1334,6 +1334,7 @@ version: 0.0.0.0
     error = _stack_version_check(repository_ctx, stack_cmd)
     if error:
         fail(error)
+    print("stack cmd:", stack_cmd)
     stack = [stack_cmd]
     exec_result = _execute_or_fail_loudly(
         repository_ctx,
@@ -1554,7 +1555,7 @@ def _download_packages_unpinned(repository_ctx, snapshot, resolved):
     remote_packages = versioned_packages + unversioned_packages
     stack = [repository_ctx.path(repository_ctx.attr.stack)]
     if remote_packages:
-        _execute_or_fail_loudly(repository_ctx, stack + ["--resolver", snapshot, "unpack"] + remote_packages)
+        _execute_or_fail_loudly(repository_ctx, stack + ["-v", "--rsl-in-log", "--resolver", snapshot, "unpack"] + remote_packages)
 
 def _invert(d):
     """Invert a dictionary."""
@@ -2380,23 +2381,28 @@ def _get_platform(repository_ctx):
 
 def _fetch_stack_impl(repository_ctx):
     repository_ctx.file("BUILD.bazel")
-    stack_cmd = repository_ctx.which("stack")
-    error = _stack_version_check(repository_ctx, stack_cmd)
-    if not error:
-        repository_ctx.symlink(stack_cmd, "stack")
-        return
-    print(error)
+    # stack_cmd = repository_ctx.which("stack")
+    # error = _stack_version_check(repository_ctx, stack_cmd)
+    # if not error:
+    #     repository_ctx.symlink(stack_cmd, "stack")
+    #     return
+    # print(error)
     print("Downloading Stack {} ...".format(_STACK_DEFAULT_VERSION))
     (os, arch) = _get_platform(repository_ctx)
     version = _STACK_DEFAULT_VERSION
     (url, sha256) = _STACK_BINDISTS[version]["{}-{}".format(os, arch)]
+    cmd = "stack"
     if "unofficial" in url:
         # the unofficial stack bindists from GHCup do not use a prefix directory
         prefix = ""
-    else:
+    elif ".tar." in url:
         prefix = paths.basename(url)[:-len(".tar.gz")]
+    else:
+        prefix = ""
+        cmd = "stack-Linux" if os.startswith("linux") else "stack-macOS"
     repository_ctx.download_and_extract(url = url, sha256 = sha256)
-    stack_cmd = repository_ctx.path(prefix).get_child("stack.exe" if os == "windows" else "stack")
+    stack_cmd = repository_ctx.path(prefix).get_child("stack.exe" if os == "windows" else cmd)
+    print(url, stack_cmd)
     if "unofficial" in url:
         # the stack binary from the unofficial bindists from GHCup lacks the executable bit
         _execute_or_fail_loudly(repository_ctx, ["chmod", "+x", stack_cmd])
